File: lib/question/ruby/namespace/symbol.rb
# frozen_string_literal: true

module Question::Ruby
  module Namespace::Symbol
    ToRecord = ->(namespace) do
      Symbol::Record.new(
        identifier: namespace.fully_qualified_name,
        kind: :namespace,
        record: namespace,
        source_locations: namespace.source_locations
      )
    end

    Index = ->(application, namespace) do
      application.symbol_index.index(ToRecord[namespace])

      namespace.children.each(&Index.curry[application])
    end
  end
end

File: web-client/src/models/symbol.ts
import { SourceLocation } from "./source_location";

export type Symbol = NamespaceSymbol | TypeInferenceSymbol;

export interface NamespaceSymbol {
  identifier: string;
  source_locations: SourceLocation[];
  kind: "namespace";
  fully_qualified_name: string;
}

export interface TypeInferenceSymbol {
  identifier: string;
  source_locations: SourceLocation[];
  kind: "type_inference";
  related_symbol_identifier: string;
}
